typed-protocols/src/Network/TypedProtocol/Pipelined.hs

```diff
@@ -44,15 +44,25 @@ data PeerSender ps (pr :: PeerRole) (st :: ps) (n :: Outstanding) c m a where
                  -> a
                  -> PeerSender ps pr st Z c m a
 
-  -- | A normal non-pipelined 'Yield'. Note that this can only be used for
-  -- messages that keep agency (or terminal messages), not for ones that
-  -- give away agency, since the 'PeerSender' has no equivalent of 'Await'.
-  -- The only way to await is via a pipelined yield: 'SenderPipeline'.
+  -- | A normal non-pipelined 'Yield'.
+  --
+  -- Note that we cannot mix pipelined and normal synchronous syle, so this
+  -- can only be used when there are no outstanding pipelined responses.
   --
   SenderYield    :: !(WeHaveAgency pr st)
-                 -> Message ps st st'
-                 -> PeerSender   ps pr (st' :: ps) n c m a
-                 -> PeerSender   ps pr (st  :: ps) n c m a
+                 -> Message    ps st st'
+                 -> PeerSender ps pr st' Z c m a
+                 -> PeerSender ps pr st  Z c m a
+
+  -- | A normal non-pipelined 'Await'. Note that this can only be used .
+  --
+  -- Note that we cannot mix pipelined and normal synchronous syle, so this
+  -- can only be used when there are no outstanding pipelined responses.
```

Comment:
Is this guaranteed by the type system?

Reply:
Ah, the `Z` in there guarantees it? I see, you have to have gone through as many `SenderCollect` branches in order to get back to `Z`. So we should have that when `n ~ Z` it's the same as a normal `Peer`.

Reply:
Yes, the `SenderYield` and `SenderAwait` both require `n ~ Z` at the type level, so that means there are no outstanding pipelined responses.

Reply:
Our replies crossed. Yes your analysis is exactly right. I'll add that to the commentary in the module.